tests/time_integration_with_constraints.cpp

```diff
@@ -0,0 +1,366 @@
+  }
+
+  invH.Invert();
+
+  FirstOrderODE ode(3);
+  ode.explicit_func = [&](const double t, const mfem::Vector& u, mfem::Vector& du_dt) {
+    mfem::Vector uf = u;
+    uf[1]           = 0;
+    du_dt           = invH * (fext(t) - K * uf - K * uc(t) - M * duc_dt(t));
+  };
+  ode.implicit_func = [&, dt_prev = -1.0, invT = mfem::DenseMatrix(3, 3)](
+                          const double t, const double dt, const mfem::Vector& u, mfem::Vector& du_dt) mutable {
+    if (dt != dt_prev) {
+      invT = M;
+      invT.AddMatrix(dt, K, 0, 0);
+      for (int i = 0; i < 3; i++) {
+        invT(i, 1) = invT(1, i) = (i == 1);
+      }
+      invT.Invert();
+    }
+
+    mfem::Vector uf     = u;
+    mfem::Vector duf_dt = du_dt;
+    duf_dt[1] = uf[1] = 0;
+
+    mfem::Vector f = fext(t) - M * duc_dt(t) - K * uc(t) - K * uf;
+    f[1]           = 0;
+
+    du_dt = invT * f;
+
+    dt_prev = dt;
+  };
+
+  ode.SetMethod< mfem::ImplicitMidpointSolver >();
+
+//  auto time_integrator = mfem::ImplicitMidpointSolver();
+//  time_integrator.Init(ode);
+
+  mfem::Vector u = uc(0);
+  double       t = 0.0;
+  for (int i = 0; i < num_steps; i++) {
+    ode.Step(u, t, dt);
+    u[1] = uc(t)[1];
+  }
+
+  mfem::Vector solution(3);
+  solution[0] = -0.1443913076373983;
+  solution[1] = 0.0;
+  solution[2] = 0.04968869236260167;
+
+  mfem::Vector error = u - solution;
+
+  std::cout << error << std::endl;
+
+  return error.Norml2();
+}
+
+double first_order_ode_test(int num_steps)
+{
+  // clang-format off
+  mfem::DenseMatrix M(3, 3);
+  M(0, 0) = 2; M(0, 1) = 1; M(0, 2) = 0;
+  M(1, 0) = 1; M(1, 1) = 4; M(1, 2) = 1;
+  M(2, 0) = 0; M(2, 1) = 1; M(2, 2) = 2;
+
+  double k = 100;
+  mfem::DenseMatrix K(3, 3);
+  K(0, 0) =  k; K(0, 1) =    -k; K(0, 2) =  0;
+  K(1, 0) = -k; K(1, 1) = 2 * k; K(1, 2) = -k;
+  K(2, 0) =  0; K(2, 1) =    -k; K(2, 2) =  k;
+  // clang-format on
+
+  double gamma = 0.5;
+  double tmax  = 1.0;
+  double dt    = tmax / num_steps;
+
+  mfem::DenseMatrix invH = M;
+  mfem::DenseMatrix invT = M;
+  invT.AddMatrix(dt * gamma, K, 0, 0);
+
+  for (int i = 0; i < 3; i++) {
+    invH(i, 1) = invH(1, i) = (i == 1);
+    invT(i, 1) = invT(1, i) = (i == 1);
+  }
+
+  invH.Invert();
+  invT.Invert();
+
+  double       t      = 0.0;
+  mfem::Vector u      = uc(t);
+  mfem::Vector du_dt  = invH * (fext(t) - K * u - M * duc_dt(t));
+  du_dt[1]            = duc_dt(t)[1];
+  mfem::Vector uf     = u;
+  mfem::Vector duf_dt = du_dt;
+  uf[1] = duf_dt[1] = 0.0;
+
+  for (int i = 0; i < num_steps; i++) {
+    t += dt;
+
+    mfem::Vector f = fext(t) - M * duc_dt(t) - K * uc(t) - K * (uf + (1.0 - gamma) * duf_dt * dt);
+    f[1]           = 0;
+
+    uf += (1.0 - gamma) * duf_dt * dt;
+    duf_dt = invT * f;
+    uf += gamma * duf_dt * dt;
+
+    u = uf + uc(t);
+  }
+
+  mfem::Vector solution(3);
+  solution[0] = -0.1443913076373983;
+  solution[1] = 0.0;
+  solution[2] = 0.04968869236260167;
+
+  mfem::Vector error = u - solution;
+
+  return error.Norml2();
```

Comment:
Can these be converted to Google Test tests that loop over a couple step values and `EXPECT_LT(error.Norml2(), threshold)` for some appropriate set of thresholds?

Reply:
Good feedback-- I expect that this test file will be deleted in the final PR, since it was mostly just a way to convince myself that the operators were written correctly.